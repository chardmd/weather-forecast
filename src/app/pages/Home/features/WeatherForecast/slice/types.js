export const APIErrorType = {
  RESPONSE_ERROR: 1,
  API_KEY_INVALID: 2,
  EMPTY_FIELD: 3,
  NO_RESULT_FOUND: 4,
  RATE_LIMIT: 5,
};
