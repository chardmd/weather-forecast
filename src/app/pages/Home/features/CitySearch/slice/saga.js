import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { selectCity } from './selectors';
import { citySearchActions as actions } from '.';
import { APIErrorType } from './types';
import { request } from '../../../../../../utils/request';
import { getBaseEndpoint } from '../../../../../../utils/endpoint';

/**
 * City locations request/response handler
 */
export function* getLocations() {
  yield delay(500);
  // Select city from store
  const city = yield select(selectCity);
  if (city.length === 0) {
    yield put(actions.locationError(APIErrorType.EMPTY_FIELD));
    return;
  }
  const requestURL = `${getBaseEndpoint('search')}&q=${city}`;
  try {
    // call our util/request. We can also use axios to substitute
    let locations = yield call(request, requestURL);
    if (locations?.length > 0) {
      //only include the locations that contains the keyword
      locations = locations.filter(({ name }) =>
        name.toLowerCase().includes(city.toLowerCase()),
      );
      //we limit the resultset to 3
      locations = locations.slice(0, 3);
      yield put(actions.locationsLoaded(locations));
    } else {
      yield put(actions.locationError(APIErrorType.NO_RESULT_FOUND));
    }
  } catch (err) {
    if (err.response?.status === 401) {
      yield put(actions.locationError(APIErrorType.API_KEY_INVALID));
    } else if (err.response?.status === 403) {
      yield put(actions.locationError(APIErrorType.RATE_LIMIT));
    } else {
      yield put(actions.locationError(APIErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* citySearchSaga() {
  yield takeLatest(actions.loadLocations.type, getLocations);
}
