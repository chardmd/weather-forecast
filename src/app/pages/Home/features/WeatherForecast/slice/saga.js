import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { selectLocationName } from './selectors';
import { weatherForecastActions as actions } from '.';
import { APIErrorType } from './types';
import { request } from '../../../../../../utils/request';
import { getBaseEndpoint } from '../../../../../../utils/endpoint';

/**
 * Weather forecast request/response handler
 */
export function* getForecast() {
  yield delay(500);
  // Select locationName from store
  const locationName = yield select(selectLocationName);
  if (locationName.length === 0) {
    yield put(actions.forecastError(APIErrorType.EMPTY_FIELD));
    return;
  }
  const FORECAST_DAYS = 3;
  const requestURL = `${getBaseEndpoint(
    'forecast',
  )}&q=${locationName}&days=${FORECAST_DAYS}`;
  try {
    const {
      current = {},
      forecast: { forecastday = [] },
    } = yield call(request, requestURL);

    if (forecastday?.length > 0) {
      yield put(actions.forecastsLoaded({ current, forecastday }));
    } else {
      yield put(actions.forecastError(APIErrorType.NO_RESULT_FOUND));
    }
  } catch (err) {
    if (err.response?.status === 401) {
      yield put(actions.forecastError(APIErrorType.API_KEY_INVALID));
    } else if (err.response?.status === 403) {
      yield put(actions.forecastError(APIErrorType.RATE_LIMIT));
    } else {
      yield put(actions.forecastError(APIErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* weatherForecastSaga() {
  yield takeLatest(actions.loadForecasts.type, getForecast);
}
