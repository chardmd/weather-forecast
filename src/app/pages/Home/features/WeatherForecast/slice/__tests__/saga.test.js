import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '..';

import { weatherForecastSaga, getForecast } from '../saga';
import { APIErrorType } from '../types';

describe('getForecast Saga', () => {
  let locationName;
  let forecastResult;
  let getForecastsIterator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getForecastsIterator = getForecast();
    const delayDescriptor = getForecastsIterator.next().value;
    expect(delayDescriptor).toMatchSnapshot();

    const selectDescriptor = getForecastsIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should return error if locationName is empty', () => {
    locationName = '';
    const putDescriptor = getForecastsIterator.next(locationName).value;
    expect(putDescriptor).toEqual(
      put(slice.weatherForecastActions.forecastError(APIErrorType.EMPTY_FIELD)),
    );

    const iteration = getForecastsIterator.next();
    expect(iteration.done).toBe(true);
  });

  it('should dispatch the forecastsLoaded action if it requests the data successfully', () => {
    locationName = 'test';
    forecastResult = {
      forecast: {
        current: {},
        forecastday: [{ name: 'test' }],
      },
    };
    const {
      forecast: { current, forecastday = [] },
    } = forecastResult;

    const requestDescriptor = getForecastsIterator.next(locationName).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getForecastsIterator.next(forecastResult).value;
    expect(putDescriptor).toEqual(
      put(
        slice.weatherForecastActions.forecastsLoaded({ current, forecastday }),
      ),
    );
  });

  it('should dispatch the city not found error', () => {
    locationName = 'test';

    const requestDescriptor = getForecastsIterator.next(locationName).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getForecastsIterator.throw({
      response: { status: 401 },
    }).value;
    expect(putDescriptor).toEqual(
      put(
        slice.weatherForecastActions.forecastError(
          APIErrorType.API_KEY_INVALID,
        ),
      ),
    );
  });
  it('should dispatch the city has no forecastResult error', () => {
    locationName = 'test';
    forecastResult = {
      forecast: {
        forecastday: [],
      },
    };

    const requestDescriptor = getForecastsIterator.next(locationName).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getForecastsIterator.next(forecastResult).value;
    expect(putDescriptor).toEqual(
      put(
        slice.weatherForecastActions.forecastError(
          APIErrorType.NO_RESULT_FOUND,
        ),
      ),
    );
  });
  it('should dispatch rate limit error', () => {
    locationName = 'test';

    const requestDescriptor = getForecastsIterator.next(locationName).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getForecastsIterator.throw({
      response: { status: 403 },
    }).value;
    expect(putDescriptor).toEqual(
      put(slice.weatherForecastActions.forecastError(APIErrorType.RATE_LIMIT)),
    );
  });

  it('should dispatch the response error', () => {
    locationName = 'test';

    const requestDescriptor = getForecastsIterator.next(locationName).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getForecastsIterator.throw(
      new Error('some error'),
    ).value;
    expect(putDescriptor).toEqual(
      put(
        slice.weatherForecastActions.forecastError(APIErrorType.RESPONSE_ERROR),
      ),
    );
  });
});

describe('weatherForecastSaga Saga', () => {
  const weatherForecastIterator = weatherForecastSaga();
  it('should start task to watch for loadForecasts action', () => {
    const takeLatestDescriptor = weatherForecastIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.weatherForecastActions.loadForecasts.type, getForecast),
    );
  });
});
