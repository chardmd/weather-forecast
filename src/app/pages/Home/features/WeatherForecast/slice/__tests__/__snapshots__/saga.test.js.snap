// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getForecast Saga should dispatch rate limit error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch rate limit error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getForecast Saga should dispatch rate limit error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/v1/forecast.json?key=undefined&q=test&days=3",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the city has no forecastResult error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the city has no forecastResult error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getForecast Saga should dispatch the city has no forecastResult error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/v1/forecast.json?key=undefined&q=test&days=3",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the city not found error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the city not found error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getForecast Saga should dispatch the city not found error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/v1/forecast.json?key=undefined&q=test&days=3",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the forecastsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the forecastsLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getForecast Saga should dispatch the forecastsLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/v1/forecast.json?key=undefined&q=test&days=3",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getForecast Saga should dispatch the response error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/v1/forecast.json?key=undefined&q=test&days=3",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should return error if locationName is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getForecast Saga should return error if locationName is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
